src/Console/CommandFactory.php

```diff
@@ -0,0 +1,39 @@
+<?php
+/**
+ * CakePHP(tm) : Rapid Development Framework (http://cakephp.org)
+ * Copyright (c) Cake Software Foundation, Inc. (http://cakefoundation.org)
+ * Licensed under The MIT License
+ * For full copyright and license information, please see the LICENSE.txt
+ * Redistributions of files must retain the above copyright notice.
+ *
+ * @copyright     Copyright (c) Cake Software Foundation, Inc. (http://cakefoundation.org)
+ * @link          http://cakephp.org CakePHP(tm) Project
+ * @license       http://www.opensource.org/licenses/mit-license.php MIT License
+ */
+namespace Cake\Console;
+
+use InvalidArgumentException;
+
+class CommandFactory implements CommandFactoryInterface
+{
+
+    /**
+     * {@inheritDoc}
+     */
+    public function create($className, ConsoleIo $io)
```

Comment:
Should we also have interface contracts for ConsoleIo as typehints? This couples it tightly to the concrete implementation.

Reply:
`Command::run` takes `ConsoleIo` as a second argument. But the class has been added in 3.6 so it's not late to change that.

Reply:
Another option would be to remove the typehints on concrete classes. I'm not sure ConsoleIo needs an interface as it isn't something that userland code should be replacing frequently.

Reply:
An interface could become handy in tests. But I think we can leave this class as it is and get back to it in 4.0 if necessary,

Reply:
Maybe `ConsoleIo $io` should be removed from a `create()` method? 

This is needed for `Shell` contructor which takes a `ConsoleIo` instance as an optional paramter. Since Shells are going to be deprecated and `$io` is an optional parameter, the proper io instance could be set by a command runner (there is `Shell::setIo()` setter). This way we wouldn't need to change the interface after shells get removed in the future.

Reply:
Ok, I'll change the interface and remove the `$io` param later this week.